src/NHibernate/Impl/AsyncEnumerableImpl.cs

```diff
@@ -0,0 +1,327 @@
+
+		internal sealed partial class AsyncEnumeratorImpl : IAsyncEnumerator<T>, IEnumerator<T>
+		{
+			private static readonly INHibernateLogger Log = NHibernateLogger.For(typeof(AsyncEnumeratorImpl));
+
+			private readonly InitializeEnumerableAsync _initializeAsync;
+			private readonly InitializeEnumerable _initialize;
+			private readonly bool _readOnly;
+			private readonly IType[] _types;
+			private readonly string[][] _columnNames;
+			private readonly RowSelection _selection;
+			private readonly IResultTransformer _resultTransformer;
+			private readonly string[] _returnAliases;
+			private readonly IEventSource _session;
+			private readonly CancellationToken _cancellationToken;
+			private readonly bool _single;
+
+			private DbDataReader _reader;
+			private DbCommand _command;
+			private bool _hasNext;
+			private bool _startedReading; // True if at least one MoveNext call was made.
+			// when we start enumerating through the DataReader we are positioned
+			// before the first record we need
+			private int _currentRow = -1;
+			private bool _isAlreadyDisposed;
+
+			public AsyncEnumeratorImpl(
+				InitializeEnumerable initialize,
+				InitializeEnumerableAsync initializeAsync,
+				bool readOnly,
+				IType[] types,
+				string[][] columnNames,
+				RowSelection selection,
+				IResultTransformer resultTransformer,
+				string[] returnAliases,
+				IEventSource session,
+				CancellationToken cancellationToken)
+			{
+				_initialize = initialize;
+				_initializeAsync = initializeAsync;
+				_readOnly = readOnly;
+				_types = types;
+				_single = types.Length == 1;
+				_columnNames = columnNames;
+				_selection = selection;
+				_resultTransformer = resultTransformer;
+				_returnAliases = returnAliases;
+				_session = session;
+				_cancellationToken = cancellationToken;
+			}
+
+			public T Current { get; private set; }
+
+			object IEnumerator.Current => Current;
+
+			public async ValueTask<bool> MoveNextAsync()
+			{
+				if (_reader == null)
+				{
+					await InitializeAsync().ConfigureAwait(false);
+				}
+
+				await ReadAsync(_cancellationToken).ConfigureAwait(false);
+
+				return _hasNext;
+			}
+
+			public bool MoveNext()
+			{
+				if (_reader == null)
+				{
+					Initialize();
+				}
+
+				Read();
+
+				return _hasNext;
+			}
+
+			private void Read()
+			{
+				try
+				{
+					_hasNext = _reader.Read();
+					_startedReading = true;
+					_currentRow++;
+				}
+				catch (DbException e)
+				{
+					throw ADOExceptionHelper.Convert(_session.Factory.SQLExceptionConverter, e, "Error executing Enumerable() query",
+													   new SqlString(_command.CommandText));
+				}
+
+				PostRead();
+			}
+
+			private void PostRead()
+			{
+				if (_selection != null && _selection.MaxRows != RowSelection.NoValue)
+				{
+					_hasNext = _hasNext && (_currentRow < _selection.MaxRows);
+				}
+
+				bool sessionDefaultReadOnlyOrig = _session.DefaultReadOnly;
+				_session.DefaultReadOnly = _readOnly;
```

Comment:
Does it really make sense to do it for each processed item? Maybe it should be done once for iterations (so this logic should be moved to `GetAsyncEnumerables` or even to `JoinedAsyncEnumerable`

Reply:
> Does it really make sense to do it for each processed item?

The user may change the `DefaultReadOnly` value while iterating, which may cause a readonly query to load entites as writable, if `DefaultReadOnly` is not set for each item.